<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>manipulation, validation and exploration of pedigrees &mdash; peddy 0.1.0 documentation</title>
    
    <link rel="stylesheet" href="_static/haiku.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="peddy 0.1.0 documentation" href="#" /> 
  </head>
  <body>
      <div class="header"><h1 class="heading"><a href="#">
          <span>peddy 0.1.0 documentation</span></a></h1>
        <h2 class="heading"><span>manipulation, validation and exploration of pedigrees</span></h2>
      </div>
      <div class="topnav">
      
        <p>
        <a class="uplink" href="#">Contents</a>
        </p>

      </div>
      <div class="content">
        
        
  <div class="section" id="manipulation-validation-and-exploration-of-pedigrees">
<h1>manipulation, validation and exploration of pedigrees<a class="headerlink" href="#manipulation-validation-and-exploration-of-pedigrees" title="Permalink to this headline">Â¶</a></h1>
<p>Pedagree is a python module for manipulating pedigree files.
While the python module has several useful features,
most users will be intereseted in <cite>peddy</cite> for its
utility in finding discrepancies between the relationships specified
in <cite>.ped</cite> files and the relationships inferred from a
corresponding <cite>.vcf</cite> file such as can occur from sample-swaps
or pedigree misspecifications.</p>
<p>The python code to do this is quite simple. But we can automate multiple
checks using the command-line interface. Below, we check for pedigree
violations by looking at 5,000 sites (see <tt class="xref doc docutils literal"><span class="pre">relatedness</span></tt>
for more details on selection) and for sex discrepancies by looking at the
non-PA regions of the X chromosome where males should have very few HET calls.</p>
<p>The command-line usage looks like:</p>
<div class="highlight-bash"><div class="highlight"><pre>python -m peddy --plot --prefix ceph-1463 ceph1463.vcf.gz ceph1463.ped
</pre></div>
</div>
<p>This will create the images:</p>
<img alt="_images/ped-check.png" src="_images/ped-check.png" />
<p>where samples are placed in the plot according to their relationship inferred from
the genotypes in the VCF and colored by their expected relationship from the PED file.
Parent-child relations (which should have IBS0 of 0) have an outline to differentiate
them from sib-sib relations.</p>
<p>and</p>
<img alt="_images/sex-check.png" src="_images/sex-check.png" />
<p>From both of these cases, we can see that it doesn&#8217;t look like there are any
sample mixups. See the docs here for an example of how a sample mixup appears.</p>
<p>For each of those images, there is a corresponding <cite>.csv</cite> file.</p>
<p>The <cite>sex-check</cite> file will look like:</p>
<div class="highlight-python"><div class="highlight"><pre>error   het_count   het_ratio   hom_alt_count   hom_ref_count   ped_sex predicted_sex   sample
FALSE   27056   0.1263089096893163  86640   127565  male    male    NA12877
FALSE   103578  0.7371260212360158  48197   92319   female  female  NA12878
FALSE   99578   0.7035623949015785  50029   91505   female  female  NA12879
FALSE   98933   0.6946225083727102  51852   90575   female  female  NA12880
FALSE   97733   0.6820833856762001  51843   91443   female  female  NA12881
FALSE   25568   0.11903775333001225 85114   129675  male    male    NA12882
FALSE   25740   0.12030398489423158 82830   131128  male    male    NA12883
FALSE   26380   0.12349609100697533 84861   128749  male    male    NA12884
FALSE   98882   0.68974609375   51608   91752   female  female  NA12885
FALSE   27341   0.12795900219965367 85473   128197  male    male    NA12886
FALSE   101838  0.7297442548709092  49321   90232   female  female  NA12887
FALSE   24317   0.11512477334381199 84141   127082  male    male    NA12888
FALSE   28980   0.13943418013856812 79538   128302  male    male    NA12889
FALSE   100617  0.7108129874533741  48475   93077   female  female  NA12890
FALSE   24443   0.11661792279543319 83839   125760  male    male    NA12891
FALSE   100453  0.7119832162677459  47684   93405   female  female  NA12892
FALSE   24774   0.11745685567987862 84482   126438  male    male    NA12893
</pre></div>
</div>
<p>And we can see that there are no inferred errors along with other information used to make that inference.</p>
<p>The <cite>ped-check</cite> file will contain <cite>n_samples</cite> * <cite>n_samples</cite> number of rows like below:</p>
<div class="highlight-python"><div class="highlight"><pre>ibs0        ibs2    ibs2*   n       rel     sample_a        sample_b        pedigree_relatedness    pedigree_parents        predicted_parents       parent_error
0.0801      0.0042  0.0165  4722    0.0245  NA12877 NA12878 0.0000  False   False   False
0.0000      0.0125  0.0360  4723    0.5401  NA12877 NA12879 0.5000  True    True    False
0.0000      0.0108  0.0390  4723    0.5482  NA12877 NA12880 0.5000  True    True    False
0.0000      0.0119  0.0407  4723    0.5770  NA12877 NA12881 0.5000  True    True    False
0.0000      0.0108  0.0351  4723    0.5448  NA12877 NA12882 0.5000  True    True    False
0.0000      0.0112  0.0392  4723    0.5438  NA12877 NA12883 0.5000  True    True    False
0.0000      0.0114  0.0379  4723    0.5452  NA12877 NA12884 0.5000  True    True    False
0.0000      0.0110  0.0358  4723    0.5341  NA12877 NA12885 0.5000  True    True    False
0.0000      0.0114  0.0390  4723    0.5597  NA12877 NA12886 0.5000  True    True    False
</pre></div>
</div>
<p>Using <cite>IBS0</cite>, we can infer the parent-child relationship of a pair of samples. If the inferred value
doesn&#8217;t match the relationship in the ped file, then the <cite>parent_error</cite> column will be <cite>True</cite>.
We can see that the first line above has a high <cite>ibs0</cite> value (and low <cite>rel</cite> value) indicating it&#8217;s
not expected to be a parent-child relation while the rest have low <cite>ibs0</cite> (and high <cite>rel</cite> values).</p>
</div>


      </div>
      <div class="bottomnav">
      
        <p>
        <a class="uplink" href="#">Contents</a>
        </p>

      </div>

    <div class="footer">
        &copy; Copyright 2015, Brent S Pedersen.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.3.
    </div>
  </body>
</html>