<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>QC &mdash; peddy 0.1.0 documentation</title>
    
    <link rel="stylesheet" href="_static/haiku.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="peddy 0.1.0 documentation" href="index.html" /> 
  </head>
  <body>
      <div class="header"><h1 class="heading"><a href="index.html">
          <span>peddy 0.1.0 documentation</span></a></h1>
        <h2 class="heading"><span>QC</span></h2>
      </div>
      <div class="topnav">
      
        <p>
        <a class="uplink" href="index.html">Contents</a>
        </p>

      </div>
      <div class="content">
        
        
  <div class="section" id="qc">
<h1>QC<a class="headerlink" href="#qc" title="Permalink to this headline">¶</a></h1>
<div class="section" id="relatedness-calculations">
<h2>relatedness calculations<a class="headerlink" href="#relatedness-calculations" title="Permalink to this headline">¶</a></h2>
<p>Using cyvcf2, we can quickly calculate relatedness using the method
described in <a class="reference external" href="http://www.nature.com/ng/journal/v42/n7/full/ng.608.html">http://www.nature.com/ng/journal/v42/n7/full/ng.608.html</a> in
equation 6.</p>
<p>This is computationally intensive to calculate for the entire genome
and other methods, such as KING (<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/20926424">http://www.ncbi.nlm.nih.gov/pubmed/20926424</a>)
have fast methods to do this on the entire genome.</p>
<p>The limitations of these methods is that they assume the average pair of samples
is unrelated.</p>
<p>In <cite>peddy</cite>, we use about 5,000 variants described in <a class="reference external" href="http://www.nature.com/nature/journal/v506/n7487/full/nature12975.html">http://www.nature.com/nature/journal/v506/n7487/full/nature12975.html</a>
that are known to be targeted by most exome platforms, in hardy weinberg equilibrium in 1000 genomes,
and mostly unlinked.</p>
<p>When a user requests to calculate relatedness, we use those 5K sites and
the genotypes from the 2504 1KG samples to provide a background of samples so
that most samples are indeed unrelated. Since we are sampling on 5K sites,
the calculations are quite fast (~5 minutes) and match very well what
we get from a whole-genome scan because of the properties of those sites.
Though we use the additional 2504 1KG samples internally, only the information
from the ped/vcf file is returned.</p>
<p><cite>peddy</cite> stores the relationships from the pedigree in a graph and can calculate
the expected coefficient of relation from what is specified in the ped file.
We can use this information to compare to what is calculated from the genotypes.</p>
</div>
<div class="section" id="sex-qc">
<h2>sex QC<a class="headerlink" href="#sex-qc" title="Permalink to this headline">¶</a></h2>
<p>We know that males should be called as either homozygous reference
or homozygous alternate on the X chromosome where as females will
have more heterozygotes. With that in mind, we can find sample swaps
that involve sex by observing the proportion of heterozygote calls.
If a sample is indicated to be male by the ped file, it should have
a low value for the proportion of het calls.</p>
</div>
<div class="section" id="het-qc">
<h2>het QC<a class="headerlink" href="#het-qc" title="Permalink to this headline">¶</a></h2>
<p>We also check that het-calls in general have an alternate count that is
about 50% of the total reads. This only makes sense for germline variant
calling but is useful for finding contamination. The actual metric is the
inter quartile range of the alternate ratio. For perfect calls, they should
all be exactly 0.5 so the range will be 0. With contamination, there will
be much more of a range around 0.5.</p>
<p>We can also check the proportion of heterozygote calls. In a contaminated
sample the number of het calls will be much higher.</p>
</div>
</div>


      </div>
      <div class="bottomnav">
      
        <p>
        <a class="uplink" href="index.html">Contents</a>
        </p>

      </div>

    <div class="footer">
        &copy; Copyright 2015, Brent S Pedersen.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.3.
    </div>
  </body>
</html>